### ISSUE TITLE
Languages bundled with releases

### ISSUE THREAD

**Issue by philwareham** [textpattern/textpattern#838]
From 4.7 onwards, we propose to bundle all Textpacks and Pophelp files into releases directly - not rely on the RPC server. To that end I have created a [new branch to track development](https://github.com/textpattern/textpattern/tree/languages-inclusion).

We will need to consider a suitable directory structure that allows us to house a textpack and pophelps per language. We could also take this opportunity to break the Textpacks into smaller chunks so the whole file is not read in entirety on every admin panel load. Since Textpacks are currently sectioned by their panel (plus `common` and `public`) this shouldn't be a problem to implement?

Directory structure proposal option 1 (more concise):

    textpattern
    -- lang
    ---- en
    ------ textpack.txt
    ------ help.xml
    ---- fr
    ------ textpack.txt
    ------ help.xml
    ---- etc..

Directory structure proposal option 2 (more granular):

    textpattern
    -- lang
    ---- en
    ------ textpack
    -------- admin.txt
    -------- article.txt
    -------- common.txt
    -------- etc...
    ------ help
    -------- admin
    ---------- about_privileges.txt
    ---------- add_new_author.txt
    ---------- etc...
    -------- article
    ---------- article_image.txt
    ---------- etc...
    ---- fr
    ------ textpack
    -------- admin.txt
    -------- article.txt
    -------- common.txt
    -------- etc...
    ------ help
    -------- admin
    ---------- about_privileges.txt
    ---------- add_new_author.txt
    ---------- etc...
    -------- article
    ---------- article_image.txt
    ---------- etc...
    -------- etc...
    ---- etc..

Could use `XML` or plain `.txt` files for Pophelps, whichever is better. If XML then we could consolidate all of a sections help into single file. I assume Textpacks would need to stay as plain `.txt` files as current format due to plugin compatibility.

Textpacks are currently being converted to raw HTML instead of Textile so they are independent of any text filter reliance. See https://github.com/textpattern/textpattern/issues/292.

Since setup language strings would be in the above, we can lose the current `setup-lang.php` file.

We discussed various ways the languages could be updated *between* releases - including checking a location/server for newer versions - but came to the conclusion that simplest route would be to clearly link to the downloadable files and document the process of manually updating. If larger translations are made between a normal point release then we will release a minor point release with those changes bundled at a logical interval. This removes the need for an RPC server.

If a better solution than the above is suggested, then we can investigate (GitHub tagging of language updated could be a possibility to allow a version update check, but since we are trying to move away from a reliance on server connection we were not particularly keen to explore that route - and also keen to avoid reliance on a third party platform).

Regarding language/country code changes, see https://github.com/textpattern/textpacks/issues/162.

Having every language already bundled into the system can potentially make per-user language preferences a reality too. See https://github.com/textpattern/textpattern/issues/644.

Once language implementation is decided, I can improve the current languages admin panel to reflect that. See https://github.com/textpattern/textpattern/issues/692.

**Comment by petecooper** [comment]
I've converted `en-gb` pophelps here (via Pandoc): https://github.com/textpattern/pophelp/commit/9c9aedd77d896462bb0312c0889971477d3e5575

Note that the converted HTML sits alongside the equivalent Textile should any sanity checking be required.

**Comment by philwareham** [comment]
OK, I've now bundled the current 4.7 language strings for all languages into the `lang` directory in the branch. They are currently in the root of that directory with filenames corresponding to their languages (such as `en.txt` and `de.txt`).

We need to agree on a directory/file structure within the `lang` directory that works and is simple for users to install updated Textpacks.

Also the format within the text files has changed slightly - there is now a `@#lang` section (which we can flesh out with strings as we work in the new panel) and there aren't currently checksums in the first line (where the version number resides). Not sure if those were ever used anyway.

So, basically, I've done all I can without any further direction from devs.

**Comment by philwareham** [comment]
This is merged now. Closing issue.
